### PR TITLE
Remove unused types when generating client filtering by tags

### PR DESCRIPTION
**Pull request by aneeshafedo**
Resolves https://github.com/ballerina-platform/ballerina-openapi/issues/405

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: openapi-cli/src/main/java/io/ballerina/openapi/cmd/CodeGenerator.java [hunk 1]
```diff
@@ -423,18 +456,109 @@ private void setGeneratedFileName(File[] listFiles, GenSrcFile gFile, int duplic
...
+            tempSourceFiles.put(TYPE_FILE_NAME, schemaContent);
+            unusedTypeDefinitionNameList = getUnusedTypeDefinitionNameList(tempSourceFiles);
+        }
+        return schemaContent;
+    }
 
-        return sourceFiles;
+    private List<String> getUnusedTypeDefinitionNameList(Map<String, String> srcFiles) throws IOException {
+        List<String> unusedTypeDefinitionNameList = new ArrayList<>();
+        Path tmpDir = Files.createTempDirectory("openapi-tmp" + System.nanoTime());
```
Comment (indikasampath2000) [inline openapi-cli/src/main/java/io/ballerina/openapi/cmd/CodeGenerator.java:505]: What if we create a folder as `.tmp`? Also, I couldn't find where you are deleting this temp directory.

File: openapi-cli/src/main/java/io/ballerina/openapi/cmd/CodeGenerator.java [hunk 2]
```diff
@@ -423,18 +456,109 @@ private void setGeneratedFileName(File[] listFiles, GenSrcFile gFile, int duplic
...
+                    for (ChildNodeEntry childNodeEntry : node.childEntries()) {
+                        if (childNodeEntry.name().equals("typeName")) {
+                            if (unusedTypeDefinitionNameList.contains(childNodeEntry.node().get().toString())) {
+                                unusedTypeDefinitionNodeList.add(node);
+                            }
+                        }
+                    }
+                } else if (node.kind().equals(SyntaxKind.ENUM_DECLARATION)) {
+                    for (ChildNodeEntry childNodeEntry : node.childEntries()) {
+                        if (childNodeEntry.name().equals("identifier")) {
```
Comment (indikasampath2000) [inline openapi-cli/src/main/java/io/ballerina/openapi/cmd/CodeGenerator.java:483]: `identifier` should be a constant.

File: openapi-cli/src/main/java/io/ballerina/openapi/cmd/CodeGenerator.java [hunk 3]
```diff
@@ -423,18 +456,109 @@ private void setGeneratedFileName(File[] listFiles, GenSrcFile gFile, int duplic
...
+        tempSourceFiles.put(TYPE_FILE_NAME, schemaContent);
+        List<String> unusedTypeDefinitionNameList = getUnusedTypeDefinitionNameList(tempSourceFiles);
+        while (unusedTypeDefinitionNameList.size() > 0) {
+            ModulePartNode modulePartNode = schemaSyntaxTree.rootNode();
+            NodeList<ModuleMemberDeclarationNode> members = modulePartNode.members();
+            List<ModuleMemberDeclarationNode> unusedTypeDefinitionNodeList = new ArrayList<>();
+            for (ModuleMemberDeclarationNode node : members) {
+                if (node.kind().equals(SyntaxKind.TYPE_DEFINITION)) {
+                    for (ChildNodeEntry childNodeEntry : node.childEntries()) {
+                        if (childNodeEntry.name().equals("typeName")) {
```
Comment (indikasampath2000) [inline openapi-cli/src/main/java/io/ballerina/openapi/cmd/CodeGenerator.java:475]: `typeName` should be a constant.

File: openapi-cli/src/main/java/io/ballerina/openapi/cmd/CodeGenerator.java (hunk 1 above)
Comment (aneeshafedo) [reply openapi-cli/src/main/java/io/ballerina/openapi/cmd/CodeGenerator.java]: Updated

File: openapi-cli/src/main/java/io/ballerina/openapi/cmd/CodeGenerator.java (hunk 2 above)
Comment (aneeshafedo) [reply openapi-cli/src/main/java/io/ballerina/openapi/cmd/CodeGenerator.java]: Added

File: openapi-cli/src/main/java/io/ballerina/openapi/cmd/CodeGenerator.java (hunk 3 above)
Comment (aneeshafedo) [reply openapi-cli/src/main/java/io/ballerina/openapi/cmd/CodeGenerator.java]: Added

**Review by indikasampath2000 — APPROVED** [review]

**Pull request merged**